src/python/pants/backend/go/util_rules/tests_analysis.py

```diff
@@ -0,0 +1,134 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import json
+import pkgutil
+from dataclasses import dataclass
+
+from pants.backend.go.util_rules.compile import CompiledGoSources, CompileGoSourcesRequest
+from pants.backend.go.util_rules.import_analysis import GatheredImports, GatherImportsRequest
+from pants.backend.go.util_rules.link import LinkedGoBinary, LinkGoBinaryRequest
+from pants.engine.fs import CreateDigest, Digest, FileContent, MergeDigests
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.util.ordered_set import FrozenOrderedSet
+
+
+@dataclass(frozen=True)
+class AnalyzeTestSourcesRequest:
+    digest: Digest
+    paths: FrozenOrderedSet[str]
+
+
+@dataclass(frozen=True)
+class GoTestCase:
+    name: str
+    package: str
+
+    @classmethod
+    def from_json_dict(cls, data: dict) -> GoTestCase:
+        return cls(name=data["name"], package=data["package"])
+
+
+@dataclass(frozen=True)
+class AnalyzedTestSources:
+    tests: FrozenOrderedSet[GoTestCase]
+    benchmarks: FrozenOrderedSet[GoTestCase]
+    has_test_main: bool
+
+    @classmethod
+    def from_json_dict(cls, data: dict) -> AnalyzedTestSources:
+        def ensure_list(xs):
+            return xs if xs is not None else []
```

Comment:
Why is this used if you're using `data.get(x, [])` already?

Reply:
Also you can use `foo or []` instead

Reply:
The Go `json` module is producing `null` values even though I included `omit_empty` in the field tags. So the key is present but is `null`.

Reply:
Will add a comment.

Reply:
Hm then how about using `foo.get(key) or []` and not using the helper function? That seems the simplest. Note that `get(key)` returns `None` if not present

Reply:
Done